### PR TITLE
- Add adjr2

### PR DESCRIPTION
**Pull request by jonpsy**
# Description

Brief Description of the PR:

Fixes # (issue)
#2350 

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [x] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [x] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [ ] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops
- [ ] 
@bhack

### CODE REVIEW

**Comment by jonpsy** [comment]
All the tests seem to be passing, would appreciate a review :)

File: tensorflow_addons/metrics/r_square.py [hunk 1]
```diff
@@ -125,18 +132,25 @@ def update_state(self, y_true, y_pred, sample_weight=None) -> None:
             tf.reduce_sum((y_true - y_pred) ** 2 * sample_weight, axis=0)
         )
         self.count.assign_add(tf.reduce_sum(sample_weight, axis=0))
+        self.num_examples.assign_add(tf.size(y_true))
 
     def result(self) -> tf.Tensor:
         mean = self.sum / self.count
         total = self.squared_sum - self.sum * mean
         raw_scores = 1 - (self.res / total)
+        n = tf.cast(self.num_examples, dtype=tf.float32)
```
Comment (bhack) [inline tensorflow_addons/metrics/r_square.py:141]: It is better that you move the code in an internal function that converts R2 in an adjusted R2 so that it is computed only if it is required the adjusted version.

**Review by bhack — CHANGES_REQUESTED** [review]

**Review by bhack — DISMISSED** [review]

**Comment by bhack** [comment]
Can you resolve conflicts?

**Comment by bhack** [comment]
Please fix tests

**Comment by jonpsy** [comment]
Done :)

**Review by bhack — APPROVED** [review]

**Comment by AakashKumarNain** [comment]
Overall LGTM but I have one doubt which I have pointed out in the code itself. Can you please clarify that?

File: tensorflow_addons/metrics/r_square.py [hunk 2]
```diff
@@ -83,6 +96,8 @@ def __init__(
         dtype: AcceptableDTypes = None,
         y_shape: Tuple[int, ...] = (),
         multioutput: str = "uniform_average",
+        penalize: bool = False,
+        num_preds: tf.int32 = 0,
```
Comment (AakashKumarNain) [inline tensorflow_addons/metrics/r_square.py:100]: Why is it set to `0`? By definition, it should be equal to number of independent variables

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:100]: Does num preds arg need to be `tf.int32`?

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:100]: Does num preds arg need to be `tf.int32`?

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:100]: By definition, yes but it would get casted anyways. My point is that this argument represents the number of independent variables used for the model. I don't know why it is set to zero. Check the `p` argument [here](https://en.wikipedia.org/wiki/Coefficient_of_determination#Adjusted_R2)

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:100]: Can you take a look at pytorch lighting https://pytorch-lightning.readthedocs.io/en/latest/_modules/pytorch_lightning/metrics/functional/r2score.html ?

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:100]: I also think that the extra checks on "num_preds" as in pytorch lighting are useful also here.

**Review by bhack — CHANGES_REQUESTED** [review]
I also think that the extra checks on "num_preds" as in pytorch lighting are useful also here.

File: tensorflow_addons/metrics/r_square.py (hunk 2 above)
Comment (jonpsy) [reply tensorflow_addons/metrics/r_square.py:100]: The num_preds should be provided by the user himself,  there's no way of knowing otherwise. I've set it to zero as default because it R2 doesn't require it, only AdjR2. The pytorch_lightning also keeps it to 0 as default.

**Comment by jonpsy** [comment]
Thanks, I forgot to address some corner cases the PyTorch codes were of great help!

**Comment by jonpsy** [comment]
@bhack, How about I remove the boolean ```penalize```, and keep only ```num_preds``` defaulting to zero, but if it has some value our result will calculate adjr2 in the result function itself?

**Comment by bhack** [comment]
If you want to do this refactor is ok

**Comment by jonpsy** [comment]
Great, I've also added checks against these corner cases.

File: tensorflow_addons/metrics/r_square.py (hunk 2 above)
Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: I agree that it should be provided by the user, that's why I am a bit hesitant for a default value for this parameter. Though that is more of a API choice.

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:89]: Now 0 has a specific semantic

**Comment by bhack** [comment]
Tests are failing

File: tensorflow_addons/metrics/r_square.py (hunk 2 above)
Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: Worst case scenario: If the user isn't aware of the default argument, then the adjusted r2 score would be much lower, which the end user wouldn't be expecting

Comment (jonpsy) [reply tensorflow_addons/metrics/r_square.py:89]: Setting ```num_preds```  to a value > 0  is the only way to use AdjR2. So, it acts as both a flag and a parameter.

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: I agree. I propose this: If the user didn't provide a value for `num_preds`, then we should raise a warning that the default value of `0` is used. This way the end user would know why the score is much lower than expected. 

This is just my suggestion. Please let me know if you have something better in mind

Comment (jonpsy) [reply tensorflow_addons/metrics/r_square.py:89]: I see your point. The API mainly calculates R2 but allows AdjR2 as an additional(optional) feature. I'm using num_preds as a flag to calculate AdjR2. Incase its default, it falls back to standard R2 :). 

Maybe my naming of ```num_preds``` is a bit off, shall I rename it to ```adjusted``` so that it's more clear that its both a flag and a parameter. Pytorch lightning does the same

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: Another suggestion: Include two parameters, a boolean `adjusted` to indicate whether to calculate adjusted r2 or not. And another `num_features` to indicate the number of independent variables with a default value of `0`. 

```
a warning shouldn't be needed because that's the expected behaviour.
```

IMO, it is needed. Until unless we tell the user that they didn't set the correct value of `num_features`, the calculation will fall back to standard r2 score (the expected behavior)

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:89]: Do we have other use case in TF Addons API where optional parameters are activated only with a dependecy on a another boolean parameter flag?

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: IIRC, `cohen's kappa` has certain bool parameters for specific use cases

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:89]: It is a different from Kappa cause It Is not only a [flag argument case] (https://ardalis.com/are-boolean-flags-on-methods-a-code-smell/) but It will has another argument that depend on the flag one.

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: Okay but is there any problem with the dependency, if at all we include it? I am looking for the `cons` basically

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:89]: It is about a potential dead param that you can change but It is inactivated when the flag is false.

But we could search if this is an anti-pattern or not.

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:89]: E.g. often just the flag parameter is evaluated as code smell

Comment (jonpsy) [reply tensorflow_addons/metrics/r_square.py:89]: I've added warnings for potential incorrect values of num_features, such that it falls back to original R2 score (and there are test to check it). 

Making num_preds both as a flag and param hits two birds with one stone doesn't it? 

Instead of setting "penalize = true" And then check if the user put "num_preds" correctly( num_preds > 0),  we can just have "adjusted" which accounts for both the validity & calculation. As bhack said, a potential dead parameter.

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: Hmm got your point. Thanks for the clarification @bhack and @jonpsy I am okay with it. Just one last thing. Can we rename `num_preds` to `num_features`?

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:89]: Yes it is `number of independent regressors` so probably we could use something better then "num_preds".

Comment (jonpsy) [reply tensorflow_addons/metrics/r_square.py:89]: Sure, I wrote ```num_preds``` because it means "number of predictors". I'm open to a better name tho, anything specific in mind?

Comment (bhack) [reply tensorflow_addons/metrics/r_square.py:89]: Sometimes is the number of indipendet regressors/predictor sometimes is "number of explanatory/indipendent variables".
So for me as you like

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: @jonpsy let us know if you have something better in mind. It's the only thing left to do and then we can merge this

Comment (jonpsy) [reply tensorflow_addons/metrics/r_square.py:89]: ```num_regressors``` has a nice ring to it :)

Comment (AakashKumarNain) [reply tensorflow_addons/metrics/r_square.py:89]: Let's go with it then

**Review by AakashKumarNain — APPROVED** [review]

**Review by AakashKumarNain — APPROVED** [review]

**Comment by AakashKumarNain** [comment]
Thanks for the changes. LGTM!

**Review by bhack — APPROVED** [review]
